config.md
# How to Configure?
`xenigma` reads configurations from ~/.config/xenigma/xenigma.conf, which is a JSON
representation of a machine.

See [test data](test-data) for examples of machines. Below is an explaination of each
component and how to specify it using JSON.

## Generating A Machine
`xenigma` provides two flags, `-generate`, and `-gen-w`, that can be used to generate
a full machine.

- `-generate` generates a machine, and encrypts a message using it, the machine itself
is not saved, so the message can't be decrypted or retrieved.

- `-gen-w` generates a machine, writes it to ~/.config/xenigma/xenigma.conf for later
usage, and uses it to encrypt a message. The machine is written before usage, so it
can be used for decryption.

## Components
### Rotors
`xenigma` allows a variable number of rotors. The number of rotors is the size of
"rotors" array.

Rotor's fields are: pathways, position, step, and cycle.

#### Pathways
Pathways are the electric connections between characters. They are represented
using a map-like 26 element array where an index and a character represent a
map pair. Keys are translated into their position in the english alphabet. For
example, if pathways[0]="c", then a is mapped to c. Arrays are chosen over maps
for pathways because ordering matters.

#### Position
Position is an integer representing the current position of the rotor, and must
be reachable from the starting position ("a").

#### Step
Step is the number of positions a rotor jumps when moving one step forward.
For example, if a rotor with position="a" and step="3" jumps once, the position
will change to "d". The default step is 1.

#### Cycle
Cycle is the number of steps needed to complete a full cycle, after which the
following rotor is shifted. For example, if a rotor with cycle=13, then it
needs to complete 13 steps for the next rotor to move one step. The default
cycle is 26.

### Reflector
Reflector is connections map, which must contain all characters in the english
alphabet, and must be symmetric. Symmetry means that if "a" is connected to "b",
then "b" must also be connected to "a".

### Plugboard
Plugboard is also a connections map similar to reflector. To keep a character
unconnected/unplugged, connect it to itself.

Run `xenigma -h` for other options.


HowToConfigure.md
# How to Configure?

`xenigma` allows for configuration of all machine's componenets using 
file **~/.config/xenigma.conf**, xenigma.conf is parsed as a normal `JSON` file.

A configuration file typically looks like the following

```json
{
    "rotors": [
        {
            "pathways": ["j", "h", "s", "e", "y", "z", "r", "k", "p", "m", "x", "i", "w", "b", "v", "f", "d", "c", "a", "t", "l", "o", "n", "g", "u", "q"],
            "position": "a",
            "step": 1,
            "cycle": 26
        },
        {
            "pathways": ["n", "c", "v", "w", "q", "t", "h", "z", "o", "m", "a", "s", "x", "r", "g", "u", "d", "i", "f", "k", "j", "b", "e", "y", "p", "l"],
            "position": "b",
            "step": 1,
            "cycle": 26
        },
        {
            "pathways": ["t", "s", "h", "m", "c", "v", "n", "y", "r", "q", "p", "e", "i", "u", "k", "z", "w", "d", "j", "a", "f", "x", "g", "b", "o", "l"],
            "position": "c",
            "step": 1,
            "cycle": 26
        }
    ],

    "reflector": {
        "connections": ["q", "y", "x", "n", "o", "r", "t", "w", "v", "p", "u", "z", "s", "d", "e", "j", "a", "f", "m", "g", "k", "i", "h", "c", "b", "l"]
    },

    "plugboard": {
        "connections": ["r", "n", "w", "q", "p", "u", "v", "o", "y", "x", "s", "t", "z", "b", "h", "e", "d", "a", "k", "l", "f", "g", "c", "j", "i", "m"]
    }
}
```

## Rotors

```json
"rotors": [
    {
        "pathways": ["j", "h", "s", "e", "y", "z", "r", "k", "p", "m", "x", "i", "w", "b", "v", "f", "d", "c", "a", "t", "l", "o", "n", "g", "u", "q"],
        "position": "a",
        "step": 1,
        "cycle": 26
    }
]
```

As said before, `xenigma` allows for **any number of rotors**.
The number of rotors is the size of *"rotors"* array in **~/.config/xenigma.conf**

Rotor's fields are *pathways*, *position*, *step*, and *cycle*.

### Pathways
Pathways are the electric connections between characters.
Pathways are represented using a 26 element array where indices represent
characters and array elements represent the character they are connected to.

*For example*, if element at index 0 is "b", then "a" (character 0) is connected
to "b".

### Position
Position is an integer which represents the current position of the rotor.
The given position must be reachable from the starting position *("a")*.

### Step
Step is the number of positions a rotor shifts when stepping once (the size of
rotor's jump.)

*For example*, if a rotor at position *"a"*, with *step = 3*, steps once,
then rotor's position changes to *"d"*. The default step size is 1.

### Cycle
Cycle is the number of rotor steps considered a full cycle, after which the
following rotor steps (is shifted.)

*For example*, if a rotor has a *cycle = 13*, then the rotor needs to complete
13 steps in order for the following rotor to step once. The default cycle size is 26.

To avoid position collisions "step \* cycle" must divide 26. Given step-cycle
combinations that don't satisfy that relation are considered wrong.

## Reflector

```json
"reflector": {
    "connections": ["q", "y", "x", "n", "o", "r", "t", "w", "v", "p", "u", "z", "s", "d", "e", "j", "a", "f", "m", "g", "k", "i", "h", "c", "b", "l"]
}
```

Reflector consists of a connections array similar to pathways with a condition
that it must be *symmetric*, meaning that if *"a"* is connected to *"b"*, then
*"b"* must also be connected to *"a"*.

## Plugboard

```json
"plugboard": {
    "connections": ["r", "n", "w", "q", "p", "u", "v", "o", "y", "x", "s", "t", "z", "b", "h", "e", "d", "a", "k", "l", "f", "g", "c", "j", "i", "m"]
}
```

Plugboard, also, consists of a connections array exactly the same as a reflector.

Plugboard's connections are required to have 26 elements,
so characters not connected to anything should be connected to themselves
(in order to not be transformed).



internal/cli/flags.go
package cli

import (
	"flag"
	"fmt"
	"log"
	"os"

	"github.com/sudo-sturbia/xenigma/v3/pkg/helper"
	"github.com/sudo-sturbia/xenigma/v3/pkg/machine"
)

// Command line flags.
var (
	write     = flag.String("write", "", "write encrypted message to file at given path")
	read      = flag.String("read", "", "encrypt contents of file at given path")
	load      = flag.String("load", "", "load and use config at given path")
	verify    = flag.String("verify", "", "verifies the correctness of config at given path")
	update    = flag.Bool("update", false, "save updated config before exiting")
	defaults  = flag.Bool("default-rotors", false, "use default values for rotor-related fields")
	shortH    = flag.Bool("h", false, "print a short help message")
	longH     = flag.Bool("help", false, "print a detailed help message")
	config    = flag.Bool("config-h", false, "print configuration help message")
	correct   = flag.Int("correct", -1, "load configs, generate a new machine if incorrect")
	generate  = flag.Int("generate", -1, "generate a machine with specified number of rotors")
	generateW = flag.Int("gen-w", -1, "generate a machine and save it's configs")
)

// correctIf handles the execution of -correct flag.
// Loads configuration at ~/.config/xenigma.conf, if configuration is
// incorrect a newly generated machine with specified number of rotors
// is saved to ~/.config/xenigma.conf and returned.
func correctIf() *machine.Machine {
	if *correct > 0 { // Load ~/.config/xenigma.conf, change if wrong
		m, err := machine.Load(*correct, true)
		if err != nil {
			if m == nil {
				log.Fatal(err)
			} else {
				fmt.Println(err)
			}
		}
		return m
	}

	return nil
}

// loadIf handles the execution of -load flag.
// Loads and returns machine at given path instead of default config.
func loadIf() *machine.Machine {
	if *load != "" { // Load machine at given path
		m, err := machine.Read(*load)
		if err != nil {
			log.Fatal(err)
		}
		return m
	}

	return nil
}

// generateIf handles the execution of -generate flag.
// Returns a newly generated machine with specified number of rotors.
func generateIf() *machine.Machine {
	if *generate > 0 { // Generate a machine to use
		return machine.Generate(*generate)
	}

	return nil
}

// generatewIf handles the execution of -gen-w flag.
// Returns a newly generated machine with specified number of rotors
// and writes config to ~/.config/xenigma.conf
func generatewIf() *machine.Machine {
	if *generateW > 0 { // Generate and write to ~/.config/xenigma.conf
		m := machine.Generate(*generateW)

		err := m.Write(os.Getenv("HOME") + "/.config/xenigma.conf")
		if err != nil {
			log.Fatal(err)
		}

		return m
	}

	return nil
}

// defaultsIf handles the execution of -default-rotors flag.
func defaultsIf(m *machine.Machine) {
	if *defaults { // Use default values for rotors
		if err := m.UseRotorsDefaults(); err != nil {
			log.Fatal(err)
		}
	}
}

// writeIf handles the execution of -write flag.
// Writes message to file if option is specified.
func writeIf(encrypted string) bool {
	if *write != "" { // Write message to file at given path
		helper.WriteStringToFile(encrypted, *write)
		return true
	}

	return false
}

// readIf handles the execution of -read flag.
// Reads contents of a file and returns a string consisting of message
// + file contents. Returns message without change if can't read.
func readIf(message string) string {
	if *read != "" { // Read contents of file at given path
		readMessage := helper.ReadStringFromFile(*read)

		switch {
		case message == "" && readMessage == "":
			log.Fatal("no message given")
		case readMessage == "":
			return message
		case message == "":
			return readMessage
		default:
			return fmt.Sprintf("%s\n%s", readMessage, message)
		}
	}

	return message
}

// updateIf checks if -update flag was specified, if so updates
// config at ~/.config/xenigma.conf
func updateIf(m *machine.Machine) {
	if *update { // Write updated configs to ~/.config/xenigma.conf
		err := m.Write(os.Getenv("HOME") + "/.config/xenigma.conf")
		if err != nil {
			log.Fatal(err)
		}
	}
}

// verifyIf  handles the execution of -verify flag.
// Verifies config at given path, prints a message accordingly and exits.
func verifyIf() {
	if *verify != "" {
		defer os.Exit(0)

		_, err := machine.Read(*verify)
		if err != nil {
			fmt.Printf("INVALID: %s\n", err.Error())
		} else {
			fmt.Println("VALID")
		}
	}
}

// helpIf handles the execution of -help, -h, or -config-h flags.
// Prints specified help message and exits.
func helpIf() {
	// Short help message
	if *shortH || (flag.NArg() == 0 && flag.NFlag() == 0) {
		defer os.Exit(0)
		fmt.Println(
			"Usage\n" +
				"    xenigma [options] <message>\n" +
				"For more details use -help.")

	}

	// Detailed help message
	if *longH {
		defer os.Exit(0)
		fmt.Println(
			"Description\n" +
				"    xenigma is a modified version of the enigma encryption machine.\n" +
				"\n" +
				"Usage\n" +
				"    xenigma [options] <message>\n" +
				"\n" +
				"Options\n" +
				"    -h                           Print a short help message.\n" +
				"\n" +
				"    -help                        Print a detailed help message.\n" +
				"\n" +
				"    -config-h                    Print a help message specifying how to\n" +
				"                                 configure a machine.\n" +
				"\n" +
				"    -verify <path>               Verify the correctness of the configuration\n" +
				"                                 at the given path.\n" +
				"\n" +
				"    -generate <numberofrotors>   Generate a machine with specified number\n" +
				"                                 of rotors and use it for encryption.\n" +
				"\n" +
				"    -gen-w <numberofrotors>      Generate a machine with specified number\n" +
				"                                 of rotors, use it for encryption, and write\n" +
				"                                 generated configs to ~/.config/xenigma.conf\n" +
				"\n" +
				"    -correct <numberofrotors>    Load ~/.config/xenigma.conf, generate a new\n" +
				"                                 machine if configs are incorrect.\n" +
				"\n" +
				"    -load <path>                 Load and use config at given path instead\n" +
				"                                 of ~/.config/xenigma.conf\n" +
				"\n" +
				"    -read <path>                 Read and encrypt contents of file at given\n" +
				"                                 path. If both -read is invoked and a message\n" +
				"                                 is given as argument, both are encrypted and\n" +
				"                                 and printed seperated by a new line.\n" +
				"\n" +
				"    -write <path>                Write encrypted message to file at given path.\n" +
				"\n" +
				"    -update                      Save updated config to ~/.config/xenigma.conf\n" +
				"                                 before exiting. Updated config is config at\n" +
				"                                 ~/.config/xenigma.conf after rotor shifting.\n" +
				"\n" +
				"    -default-rotors              Use default values for rotor-related fields.\n" +
				"                                 Default values are \"a\"'s for rotor positions,\n" +
				"                                 1 for step size, and 26 for cycle size.\n" +
				"\n" +
				"xenigma is licensed under MIT license.\n" +
				"For source code check the github repo [github.com/sudo-sturbia/xenigma].")
	}

	// Configuration help message
	if *config {
		defer os.Exit(0)
		fmt.Println(
			"Configuration\n" +
				"\n" +
				"    xenigma allows for configuration of all machine's componenets through\n" +
				"    JSON. Configurations file should be located at ~/.config/xenigma.conf\n" +
				"\n" +
				"    An example of a ~/.config/xenigma.conf is the following\n" +
				"\n" +
				"    {\n" +
				"        \"rotors\": [\n" +
				"            {\n" +
				"                \"pathways\": [\"a\", \"b\", \"c\", ...],\n" +
				"                \"position\": \"a\",\n" +
				"                \"step\": 1,\n" +
				"                \"cycle\": 26\n" +
				"            },\n" +
				"            {\n" +
				"                \"pathways\": [\"a\", \"b\", \"c\", ...],\n" +
				"                \"position\": \"b\",\n" +
				"                \"step\": 1,\n" +
				"                \"cycle\": 26\n" +
				"            },\n" +
				"            {\n" +
				"                \"pathways\": [\"a\", \"b\", \"c\", ...],\n" +
				"                \"position\": \"c\",\n" +
				"                \"step\": 1,\n" +
				"                \"cycle\": 26\n" +
				"            }\n" +
				"        ],\n" +
				"        \n" +
				"        \"reflector\": {\n" +
				"            \"connections\": [\"a\", \"b\", \"c\", ...]\n" +
				"        },\n" +
				"\n" +
				"        \"plugboard\": {\n" +
				"            \"connections\": [\"a\", \"b\", \"c\", ...]\n" +
				"        }\n" +
				"    }\n" +
				"\n" +
				"    Rotors\n" +
				"\n" +
				"    xenigma allows for any number of rotors. The number of rotors is the size\n" +
				"    of \"rotors\" array in ~/.config/xenigma.conf\n" +
				"\n" +
				"    Rotor's fields are: pathways, position, step, and cycle.\n" +
				"\n" +
				"    Pathways are represented using a 26 element array where indices represent \n" +
				"    characters and array elements represent the character they are connected\n" +
				"    to. For example if element at index 0 is \"b\", then \"a\" (character 0) is \n" +
				"    connected to \"b\".\n" +
				"\n" +
				"    Position is an integer which represents the current position of the rotor, \n" +
				"    and must be reachable from the starting position (\"a\").\n" +
				"\n" +
				"    Step is the number of positions a rotor shifts when stepping once (the size \n" +
				"    of rotor's jump.) For example if a rotor at position \"a\", with step = 3, \n" +
				"    steps once, then rotor's position changes to \"d\". The default step size is 1.\n" +
				"\n" +
				"    Cycle is the number of rotor steps considered a full cycle, after which the\n" +
				"    following rotor steps (is shifted.) For example, if a rotor has a cycle = 13,\n" +
				"    then the rotor needs to complete 13 steps in order for the following rotor \n" +
				"    to step once. The default cycle size is 26.\n" +
				"\n" +
				"    Reflector\n" +
				"\n" +
				"    Reflector consists of a connections array similar to pathways with a \n" +
				"    condition that it must be symmetric, meaning that if \"a\" is connected to \"b\", \n" +
				"    then \"b\" must also be connected to \"a\".\n" +
				"\n" +
				"    Plugboard\n" +
				"\n" +
				"    Plugboard, also, consists of a connections array exactly the same as a \n" +
				"    reflector.\n" +
				"\n" +
				"    Plugboard's connections are required to have 26 elements, so characters \n" +
				"    not connected to anything should be connected to themselves (in order \n" +
				"    to not be transformed).")
	}
}
